Make install should always use phony targets

The makefile currently specifies files for installation as individual "normal" targets, for example:

```make
$(DESTDIR)$(bindir)/luarocks: $(builddir)/luarocks
        mkdir -p "$(@D)"
        $(INSTALL) "$<" "$@"
```

...but there's a reason almost no one uses normal targets for installation like this -- it fails to install if the target is newer than the dependency (which can and does happen for various reasons) and also becomes extremely buggy if [`.DELETE_ON_ERROR`](https://www.gnu.org/software/make/manual/html_node/Special-Targets.html) is ever enabled (which the `make` manual [recommends](https://www.gnu.org/software/make/manual/html_node/Errors.html)).

The only correct way to install built files is to use one or more phony targets that run the commands one-by-one _unconditionally_, e.g.:

```make
install: all
	mkdir -p '$(DESTDIR)$(bindir)/'
	$(INSTALL) '$(builddir)/luarocks' '$(DESTDIR)$(bindir)/luarocks'
	# etc....

.PHONY: install
```

Does not work:
```
Please run the ./configure script before building.

make[1]: *** No rule to make target 'config.unix', needed by 'luarocks'.  Stop.
```

--- a/GNUmakefile
+++ b/GNUmakefile
@@ -104,52 +104,50 @@ $(buildbinarydir)/luarocks-admin.exe: src/bin/luarocks-admin $(LUAROCKS_FILES)
 # Regular install
 # ----------------------------------------
 
-INSTALL_FILES = $(DESTDIR)$(bindir)/luarocks \
-	$(DESTDIR)$(bindir)/luarocks-admin \
-	$(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua \
-	$(patsubst src/%, $(DESTDIR)$(luadir)/%, $(LUAROCKS_FILES))
+INSTALL_FILES =
 
-install: $(INSTALL_FILES)
+install: all install-config
+	mkdir -p '$(DESTDIR)$(bindir)/'
+	$(INSTALL) '$(builddir)/luarocks' '$(DESTDIR)$(bindir)/luarocks'
+	$(INSTALL) '$(builddir)/luarocks-admin' '$(DESTDIR)$(bindir)/luarocks-admin'
+	find src/luarocks/ -type d | while read f; \
+	do \
+	   mkdir -p '$(DESTDIR)$(luadir)'/`echo $$f | sed 's,^src/,,'`; \
+	done
+	find src/luarocks/ -type f -name '*.lua' | while read f; \
+	do \
+	   $(INSTALL_DATA) "$$f" '$(DESTDIR)$(luadir)'/`echo $$f | sed 's,^src/,,'`; \
+	done
 
-install-config: $(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua
-
-$(DESTDIR)$(bindir)/luarocks: $(builddir)/luarocks
-	mkdir -p "$(@D)"
-	$(INSTALL) "$<" "$@"
-
-$(DESTDIR)$(bindir)/luarocks-admin: $(builddir)/luarocks-admin
-	mkdir -p "$(@D)"
-	$(INSTALL) "$<" "$@"
-
-$(DESTDIR)$(luadir)/luarocks/%.lua: src/luarocks/%.lua
-	mkdir -p "$(@D)"
-	$(INSTALL_DATA) "$<" "$@"
-
-$(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua: $(builddir)/config-$(LUA_VERSION).lua
-	mkdir -p "$(@D)"
-	$(INSTALL_DATA) "$<" "$@"
+install-config:
+	mkdir -p '$(DESTDIR)$(luarocksconfdir)/'
+	$(INSTALL_DATA)  '$(builddir)/config-$(LUA_VERSION).lua' '$(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua'
 
 uninstall:
-	rm -rf $(INSTALL_FILES)
+	rm -rf $(DESTDIR)$(bindir)/luarocks \
+	       $(DESTDIR)$(bindir)/luarocks-admin \
+	       $(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua \
+	       $(patsubst src/%, $(DESTDIR)$(luadir)/%, $(LUAROCKS_FILES))
 
 # ----------------------------------------
 # Binary install
 # ----------------------------------------
 
-LUAROCKS_CORE_FILES = $(wildcard src/luarocks/core/* src/luarocks/loader.lua)
-INSTALL_BINARY_FILES = $(patsubst src/%, $(DESTDIR)$(luadir)/%, $(LUAROCKS_CORE_FILES)) \
-	$(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua
-
-install-binary: $(INSTALL_BINARY_FILES)
+install-binary: binary install-config
 	mkdir -p "$(buildbinarydir)"
 	$(INSTALL) "$(buildbinarydir)/luarocks.exe" "$(DESTDIR)$(bindir)/luarocks"
 	$(INSTALL) "$(buildbinarydir)/luarocks-admin.exe" "$(DESTDIR)$(bindir)/luarocks-admin"
+	mkdir -p '$(DESTDIR)$(luadir)/luarocks/core'
+	for f in src/luarocks/core/*.lua src/luarocks/loader.lua; \
+	do \
+	   $(INSTALL_DATA) "$$f" '$(DESTDIR)$(luadir)'/`echo $$f | sed 's,^src/,,'`; \
+	done
 
 # ----------------------------------------
 # Bootstrap install
 # ----------------------------------------
 
-bootstrap: luarocks $(DESTDIR)$(luarocksconfdir)/config-$(LUA_VERSION).lua
+bootstrap: luarocks install-config
 	./luarocks make --tree="$(DESTDIR)$(rocks_tree)"
 
 # ----------------------------------------
@@ -185,4 +183,4 @@ clean: windows-clean
 		./.luarocks \
 		./lua_modules
 
-.PHONY: all build install binary install-binary bootstrap clean windows-binary windows-clean
+.PHONY: all build install install-config binary install-binary bootstrap clean windows-binary windows-clean

